### ISSUE TITLE
Nag messages for errors loading other mod dlls

### ISSUE THREAD

**Issue by hvacengi** [KSP-KOS/KOS#1718]
_Definitions:_
_Addon: a dll used to add a feature to kOS, usually for interoperability with other mods_
_Mod: a dll or parts pack used to add features to KSP_

As originally written, the `AssemblyWalkAttribute` code adds a nag message whenever it finds a dll that throws an error when attempting to walk the types and attributes.  This was intentional, as it would force addon authors to see that their new dll had an issue and fix it, by preventing further input at the terminal.  In the release candidate we have found that an unintended consequence is that kOS adds these nag messages for other mods as well.  Some of the mods at issue appear to not be updated for the most recent version of KSP, though others are current releases that appear to work just fine other than creating an error for kOS.  

We could spend a lot of time trying to figure out why there are issues loading these other mods, but it doesn't really change the issue: it is possible for kOS to lock out the user due to another mod.  If it was only true of "broken" mods, we might consider instructing users to remove the offending mods.  But because "working" mods can also lock out the user, we should re-evaluate the lock out strategy.  Since I came up with the original strategy on my own, I thought it might be a good idea to have a discussion about it before committing to the next strategy.

I see the following options:
1. **Only add the nag messages when compiled with the DEBUG flag.**  We would then recommend that addon authors test their addon against the debug build instead of release.  This prevents the release version from ever showing assembly load errors to the end user.  That's a good thing in terms of playing nice with other mods, but could lead to confusion if an addon ends up with the error and the user doesn't recognize what is wrong.  Since it's already important to request the log file when troubleshooting errors, I don't see this as a big limitation because we will see the load errors in the log still.
2. **Set the nag messages to display only once.**  Nag message persistence is "opt-in".  If we display the message only once, it would be enough to show the error to addon authors and also to annoy users enough to maybe comment on the issue to the other mod author.  But having our users bug the other mod authors isn't exactly a way to ingratiate ourselves the them.  The user would only need to power cycle the first cpu that loads for the KSP session.
3. **Provide some other means of displaying the error to users.**  I've toyed with the idea of providing a kOS specific debug window.  RT has a similar window available.  On a kOS specific error the window could pop up (optionally by config value) so that a user can see the detailed message without all of the other KSP messages.  The biggest down side of this option is that it would probably delay our release.  So it may be better to choose option 1 for the v1.0.0 release, and then strive for option 3 in the next release.  It may serve as a good proof of concept for our first use of the new Unity 5 GUI system.

I'd love to get thoughts on the issue, or alternative solutions.

**Comment by Dunbaratu** [comment]
I don't understand the description of the problem yet because it seems to be describing something impossible.  You're talking about DLLs that KSP loads just fine, but kOS can't walk through.  How is that even possible?  What is kOS doing with these DLLs that KSP isn't?  Can you elaborate on what's happening here without me having to go walk the code to find the answer?

**Comment by hvacengi** [comment]
I'm honestly not sure why we're getting an error either.  In all of my previous testing, the only time I get the nag message was when KSP was unable to load the dll as well.  Here's a snippet showing KSP loading a DLL without issue:

**Comment by Dunbaratu** [comment]
> when KSP was unable to load the dll as well

This phrase implies we're going out to the filesystem, to the DLL files, and repeating the work of loading them that KSP already does.  Is it not possible to walk them after KSP did the work of loading them into memory, thus meaning we don't duplicate the work and possibly load them differently than KSP is?

**Comment by hvacengi** [comment]
(I updated the above post which posted in error)

I'm using "currently loaded assemblies" this way:
```
var assemblies = AppDomain.CurrentDomain.GetAssemblies();
```
KSP does some additional fancy stuff.  Essentially every dll get's loaded into the AppDomain, and it isn't until you attempt to access them that you get many of these errors.  KSP uses a proxy Domain to load the assembly, then seems to do some magic to access them using the primary AppDomain.  Then they have some more complicated loading logic that checks for dependencies and the like to make sure they load dll's in order.  

As far as I could tell, once KSP loaded the dll we were pretty safe.  I attempted to confirm this using a dll that would load out of order when scanned automatically, but in order using KSP's dependency model.  Without the dependency declared the out of order dll's failed to load, but with the dependency declared the loaded without issue.

**Comment by hvacengi** [comment]
> it not possible to walk them after KSP did the work of loading them into memory, thus meaning we don't duplicate the work and possibly load them differently than KSP is?
It is possible, but more complicated.  We cannot do it from `kOS.Safe` because it requires a reference to KSP to make it work.  We could possibly modify `AssemblyWalkAttribute` such that we pass it a list of assemblies in the bootstrapper instead of having it create the list itself.  I can try that and see if it changes the behavior.

**Comment by dewiniaid** [comment]
Digging through the [FloaterTools source](https://github.com/BobPalmer/SrvPack/tree/0.5.4.0/FloaterTools/FloaterTools/FloaterTools) (which is pretty small), I can't really see any particular reason this would fail either.  The attribute in question is  used to define extension methods, but if that was really the case kOS would be failing on just about every Mod DLL in existence -- and it doesn't.

Digging through AssemblyWalk (but probably using the wrong branch), I can't see anything about its way of evaluating assemblies that's any different from the way other mods do it; I know offhand both KRPC and KIPC walk over assemblies using `AppDomain.CurrentDomain.GetAssemblies()` and scan for attributes.  I did notice that your version is using the `inherit = true` flag though, which I don't recall seeing in either other implementation.

**Comment by hvacengi** [comment]
I have tracked down the problem, and am preparing a different fix than those listed above.

#The issue:
FloaterTools is compiled against .net 4.5 instead of 3.5 or 2.0.  As far as I'm aware, Mono is essentially compatible with 2.0 and 3.5, and has some features of 4.0 and 4.5.  But the big issue is that 2.0, 3.0, and 3.5 share the same underlying CLR while 4.0 and 4.5 use a newer version of the CLR.  This is why it was a built in type that threw the error, because kOS expected the early CLR version, while FloaterTools was using the newer CLR version.  I'm not sure if this is actually an issue for KSP or not, but it obviously was an issue for kOS.

#The change:
@Dunbaratu and I discussed the issue pretty thuroughly and decided to introduce a new fourth option.  kOS will only attempt to load assemblies which define kOS as a dependency using the `KSPDependency` attribute, or have the dll file name start with "kOS".  That means that we're only checking against assemblies that have specifically opted to relate themselves to kOS, rather than all mods.  It also means we can keep the nag messages, because they truly mean that a mod that's interacting with kOS cannot be loaded and should be repaired/removed/replaced.

We will also update the documentation for addons instructing authors that it is important to make sure that the addon is compiled against the same version of .net that kOS is compiled against.

**Comment by dewiniaid** [comment]
> kOS will only attempt to load assemblies which define kOS as a dependency using the KSPDependency attribute, or have the dll file name start with "kOS". That means that we're only checking against assemblies that have specifically opted to relate themselves to kOS, rather than all mods.

I'd say limit it solely to the KSPAssemblyDependency attribute on the unlikely case that someone unintentionally happens to happens to prefix with "KOS", e.g. Kosmos, Kosmonaut, Kosmic.  There's no existing mods, so imposing that limitation is probably reasonable.

One possible issue just occurred to me is similar to a problem I've had in KRPC, namely: It is _extremely_ difficult to do optional dependencies using attributes since you can't add attributes at runtime.  More about this can be found [here](https://github.com/krpc/krpc/issues/143) and [here](http://forum.kerbalspaceprogram.com/index.php?/topic/130742-113-krpc-control-the-game-using-c-c-java-lua-python-ruby-haskell-v035-25th-june-2016/&do=findComment&comment=2669254).  The short version is then: Attributes are convenient, but it would probably be worthwhile to be able to accomplish the same things through a method call

**Comment by hvacengi** [comment]
The problem with only using the dependency attribute is that it prevents us from being able to use the same system internally (for kOS.dll and kOS.Safe.dll) or for anyone else who might have an option that doesn't require a reference to KSP's dlls.  Maybe I'll change it to check for "kOS." (including the period) to be a little more specific.  It is case sensitive, so the reversed capitalization would help with false positives too.

As for the lack of runtime flexibility, I honestly consider that a good thing.  Especially given KSP's loading model.  If you define `[assembly: KSPAssemblyDependency("kOS", 0, 20)]` in `Assembly.cs`, KSP won't load and use that dll unless it has already loaded a dll defining `[assembly: KSPAssembly("kOS", 0, 20)]`.  That means that, for example, Infernal Robotics could package their base dll file, along with a 2nd dll file `MagicSmokeIndustries.InfernalRobotics.kOS.dll` which would only be loaded by KSP when kOS is also installed.  Meaning that the Infernal Robotics folks wouldn't need to make the dll a separate download.

I particularly like the attribute system, because the implementation allows us to do something that C# does not support: enforce statically defined methods.  The quick and easy answer is (and this is how it was previously implemented) to simply load any class that inherited from `Addon` as an `Addon`.  The same thing with bindings and functions.  The trick was that it meant a reflection walk every single time you created a new instance of the manager class (or some caching of the value).  It worked because kOS knew how to access and cast each of these types.  This either locked us in to inheritance, or some way of enforcing a static method reference for registering types to be managed.  Using the attributes, we have offloaded all of that overhead into the more appropriate loading screen, instead of at vessel launch, and we've created a way for classes to tell us where to find the static methods, rather than attempt to enforce a language feature that doesn't exist in C#.  It isn't the `Addon` itself that drove this architecture, but rather the manager classes.

That being said, the system is flexible.  If you would like to opt in by calling a method instead of defining an attribute, the `RegisterMethod` of the addon manager is publicly accessible.  You'll have to find a way to instantiate your class first (like using a `KSPAddon` and then make the call:
```
AddonManager.RegisterMethod(new kOSAddonAttribute("name"), typeof(MyClass));
```
However this will not be the recommended method of registering.

Once of the nicer things about how I implemented that attribute system is that the attributes are not repeatedly loaded off of the class definition itself.  So the attribute really only needs to be available when the type is registered with the manager class.  After that, the manager class keeps it's own reference.  So you don't need to worry about kOS trying to go fetch the attribute information off of the class again, making dynamically registering a possibility.  

Now, if you want to create your own manager class that uses our system, it will need the `AssemblyWalkAttribute` defined.  But if you're doing your own manager, it's only one additional step to register types for the manager yourself.

**Comment by dewiniaid** [comment]
Fired up my full (not-stripped-down-for-development) copy of KSP for the first time in awhile so thought I'd mention a few more affected mods in case it helps:

```
Error while loading assembly: FloaterTools, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null, skipping assembly.
Full path: C:\Steam\SteamApps\common\Kerbal Space Program\GameData\UmbraSpaceIndustries\SrvPack\FloaterTools.dll
Error while loading assembly: IRSurfaceSampler, Version=1.0.1.4, Culture=neutral, PublicKeyToken=null, skipping assembly.
Full path: C:\Steam\SteamApps\common\Kerbal Space Program\GameData\MagicSmokeIndustries\IRSurfaceSampler\Plugin\IRSurfaceSampler.dll
Error while loading assembly: KF_plugin, Version=1.9.5730.28121, Culture=neutral, PublicKeyToken=null, skipping assembly.
Full path: C:\Steam\SteamApps\common\Kerbal Space Program\GameData\KerbalFoundries\Plugins\KF_plugin.dll
Error while loading assembly: IndicatorLights, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null, skipping assembly.
Full path: C:\Steam\SteamApps\common\Kerbal Space Program\GameData\IndicatorLights\IndicatorLights.dll
Error while loading assembly: BetterBurnTime, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null, skipping assembly.
Full path: C:\Steam\SteamApps\common\Kerbal Space Program\GameData\BetterBurnTime\BetterBurnTime.dll
```

**Comment by hvacengi** [comment]
Those error messages don't look like those used by kOS.  Can you provide the whole log, or verify that the messages come between log entries of:
```
[LOG 15:17:43.349] kOS: AssemblyWalkAttribute begin walking
```
and
```
[LOG 15:17:45.002] kOS: Finish walking assemblies.
```
I still expect similar messages to appear in the kOS loading section, but I suspect that those mods are throwing errors when KSP tries to load them.

I left off a very important part of "the solution" above: we will also only be scanning dll files that KSP reports as successfully loading itself.  I'll edit the post to reflect this.

**Comment by dewiniaid** [comment]
They were definitely from kOS; that was copy-pasted from the terminal.  That said, it looks like two of those (KF_Plugin and IRSurfaceSampler) are also failing in KSP proper now that I've looked at the log more in depth; the other tw oare fine.

Interestingly enough, kOS errors out _twice_ on each assembly though only the first was reported:

```
(line 3088)
kOS: Error while loading assembly: FloaterTools, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null, skipping assembly.
Full path: C:\Steam\SteamApps\common\Kerbal Space Program\GameData\UmbraSpaceIndustries\SrvPack\FloaterTools.dll

[...]

(Line 3129)
kOS: Error while loading assembly: FloaterTools, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null, skipping assembly.
Full path: C:\Steam\SteamApps\common\Kerbal Space Program\GameData\UmbraSpaceIndustries\SrvPack\FloaterTools.dll
 
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/UnityEngineDebugBindings.gen.cpp Line: 64)

kOS: Exception: Could not load type 'System.Runtime.CompilerServices.ExtensionAttribute' from assembly 'FloaterTools'.
Stack Trace:
  at (wrapper managed-to-native) System.MonoCustomAttrs:GetCustomAttributesInternal (System.Reflection.ICustomAttributeProvider,System.Type,bool)

  at System.MonoCustomAttrs.GetCustomAttributesBase (ICustomAttributeProvider obj, System.Type attributeType) [0x00000] in <filename unknown>:0 

  at System.MonoCustomAttrs.GetCustomAttributes (ICustomAttributeProvider obj, System.Type attributeType, Boolean inherit) [0x00000] in <filename unknown>:0 

  at System.MonoType.GetCustomAttributes (System.Type attributeType, Boolean inherit) [0x00000] in <filename unknown>:0 

  at kOS.Safe.Utilities.AssemblyWalkAttribute.WalkAssembly (System.Reflection.Assembly assembly) [0x00000] in <filename unknown>:0 

  at kOS.Safe.Utilities.AssemblyWalkAttribute.WalkAllAssemblies () [0x00000] in <filename unknown>:0 
 ```

Full log attached.

[output_log.zip](https://github.com/KSP-KOS/KOS/files/361578/output_log.zip)

**Comment by hvacengi** [comment]
Ah, I thought you were posting the log content itself and didn't realize you were posting the nag messages.

**Comment by hvacengi** [comment]
All three of the mods that aren't throwing errors outright in KSP are compiled against a version of .net not supported by Unity.

FloaterTools (USI Survival Pack): https://github.com/BobPalmer/SrvPack/blob/master/FloaterTools/FloaterTools/FloaterTools/FloaterTools.csproj#L12

Better burn time: https://github.com/KSPSnark/BetterBurnTime/blob/master/src/BetterBurnTime.csproj#L12

Indicator lights: https://github.com/KSPSnark/IndicatorLights/blob/master/src/IndicatorLights.csproj#L12

As I said above, I don't know that this is necessarily considered an issue for KSP, but the official line I've seen from unity is that .net 2.0/3.5 is the supported API framework.  As I understand it, there are also some advanced features that have been back ported into their own implementation of mono, but I could not readily find such a list.

Those details don't really affect us anymore anyways, since I will update the code not to load these other mods.  If an addon starts with "kOS." or it defines "kOS" as a dependency, it will need to be compiled against the correct .net framework to ensure that it loads.

**Comment by dewiniaid** [comment]
Looking forward to it.  If you need any early testing, point me at a branch/tag/commit-ish and I can compile at home.

**Comment by kosstarter** [comment]
I'm just starting with kOS and already running into trouble. I might be wrong but i'm afraid there might be more to it than it appears. 
After the problems as decribed in https://www.reddit.com/r/Kos/comments/4snkub/what_mods_are_incompatible_with_kos/  i did a complete new install of KSP. First addition was kOS. With CKAN i added several mods, every now and then, i checked if the terminal window appaered (which it did).
After adding 26 mod's in total, i started programming for kOS. It worked fabulous!
This morning i tried to have the terminal window open automatically whenever a craft is launched, using a boot.ks i created. I used two hints i found in REDDIT: 
CORE:PART:GETMODULE("kOSProcessor"):DOEVENT("Open Terminal").
and
core:doevent("Open Terminal").
The result was the same as i described in the REDDIT thread. I removed the boot.ks script and (after starting KSP again) tried to open the terminal window. No matter what i try no terminal window appears. By the way, i'm not using the three mentioned mods.

You can find my KSP.log file here: https://dl.dropboxusercontent.com/u/14923883/KSP.log

**Comment by hvacengi** [comment]
@kosstarter It looks like I led you astray on the reddit post.  I assumed that you had the same problem as posted here, but it does not sound like you are seeing any nag messages and you don't appear to be running the release candidate.  Could you do me a favor and create a new issue for this instead, so we can discuss the details there?
